features/user-profile.md

```diff
@@ -0,0 +1,83 @@
+# API Design Overview
+
+This document introduces the new Skygear User Record (Profile), which is
+re-designed from user record and user object, utilizing user profile discovery.
+The design depends on [the field-based ACL](field-based-acl.md) feature.
+
+## New User Record
+
+The newly designed user record contains discoverable information of user, like
+username, email, etc., and queryable information, like profile photo, status,
+etc. For discoverable information, the field-based ACL should be set as
+discoverable to any user and updatable by owner. For queryable information, the
+ACL should be set as queryable by any user and updatable by owner.
+
+Using field-based ACL, user record can serve as both public profile and private
+profile.
+
+The following table gives an example of the field-based ACL of user record
+fulfilling: 1) both username and email are discoverable; 2) status and profile
+photo are readable; 3) game score is queryable; 4) birthday is private:
+
+| Class | UserRole |    Field    |     Level    |
+|-------|----------|-------------|--------------|
+| *     | AnyUser  | *           | ReadWrite    |
+| User  | Owner    | *           | ReadWrite    |
+| User  | AnyUser  | username    | Discoverable |
+| User  | AnyUser  | email       | Discoverable |
+| User  | AnyUser  | status      | ReadOnly     |
+| User  | AnyUser  | profile_pic | ReadOnly     |
+| User  | AnyUser  | game_score  | Queryable    |
+| User  | AnyUser  | birthday    | NoAccess     |
+
+## Authentication Object
+
+The new authentication object (i.e. `_auth` object) is to replace the old
+ `_user` object, to avoid ambiguity with the user record.
+
+The authentication object contains the following fields:
+
+- `id (text)`: should be the same as `user._id` as a convention
+- `password (text)`: the password for authentication
+- `provider_info (jsonb)`: the authentication information from Authentication
+  Provider
+- `token_valid_since`: the timestamp limit the validity of the authentication
+  token
+- `last_login_at`: the timestamp when the user last login
+- `last_seen_at`: the timestamp when user last perform an action
+
+## New Sign Up Method
+
+<!-- TODO: New sign up method to help setting up public / private profile -->
+
+# Sample Codes for Use Cases
+
+<!--
+  TODO: Give some sample code demonstrate how field-based ACL is set on
+        user record
+-->
+
+<!--
+  TODO: Give some sample code demonstrate how information in user record can
+        be discovered.
+-->
+
+# Changes on SDK
+
+The SDKs would be expected to have the following changes:
+
+- All log in / sign up method would returns **a user record**, instead of
+  a `_user` object.
+- remove `getUserByEmails()`, `getUserByUsernames()`, `discoverUserByEmails()`
```

Comment:
Suggestion: Since user discovery by email and and username appears to be common use case, I think we should retain these two function. Alternative is to create helper functions that return a query object that returns profile when executed.